pennylane/decomposition/controlled_decomposition.py

```diff
@@ -0,0 +1,239 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains special logic of decomposing controlled operations."""
+
+from __future__ import annotations
+
+import functools
+from typing import Callable, Type
+
+import pennylane as qml
+from pennylane.operation import Operator
+
+from .decomposition_rule import DecompositionRule, register_resources
+from .resources import controlled_resource_rep, resource_rep
+
+
+class CustomControlledDecomposition(DecompositionRule):  # pylint: disable=too-few-public-methods
+    """A decomposition rule applicable to an operator with a custom controlled decomposition."""
+
+    def __init__(self, custom_op_type: Type[Operator]):
+        self.custom_op_type = custom_op_type
+        super().__init__(self._get_impl(), self._get_resource_fn())
+
+    def _get_impl(self):
+        """The implementation of a controlled op that decomposes to a custom controlled op."""
+
+        def _impl(*params, wires, control_wires, control_values, **_):
+            wires_0_ctrl_vals = [w for w, val in zip(control_wires, control_values) if not val]
+            for w in wires_0_ctrl_vals:
+                qml.PauliX(w)
+            self.custom_op_type(*params, wires=wires)
+            for w in wires_0_ctrl_vals:
+                qml.PauliX(w)
+
+        return _impl
+
+    def _get_resource_fn(self) -> Callable:
+        """The resource function."""
+
+        def _resource_fn(*_, num_zero_control_values, **__):
+            return {
+                resource_rep(self.custom_op_type): 1,
+                resource_rep(qml.PauliX): num_zero_control_values * 2,
+            }
+
+        return _resource_fn
+
+
+def _controlled_resource_rep(base_op_rep, num_control_wires, num_work_wires):
+    """Returns the resource rep of a controlled op, dispatches to a custom op if possible."""
+
+    if (base_op_rep.op_type, num_control_wires) in base_to_custom_ctrl_op():
+        return resource_rep(base_to_custom_ctrl_op()[(base_op_rep.op_type, num_control_wires)])
+
+    if base_op_rep.op_type in (qml.X, qml.CNOT, qml.Toffoli, qml.MultiControlledX):
+        # First call controlled_resource_rep to get flatten any nested structures
+        rep = controlled_resource_rep(
+            base_class=base_op_rep.op_type,
+            base_params=base_op_rep.params,
+            num_control_wires=num_control_wires,
+            num_zero_control_values=0,
```

Comment:
```suggestion
```

Reply:
I made a couple of suggestions where one could skip the `num_zero_control_values=0` due to the new default value. They are of course optional and if you prefer to keep them for readability or so, happy to keep them